### ISSUE TITLE
Can

### ISSUE THREAD

**Issue by dathudeptrai** [tensorflow/tensorflow#41760]
<em>Please make sure that this is a bug. As per our
[GitHub Policy](https://github.com/tensorflow/tensorflow/blob/master/ISSUES.md),
we only address code/doc bugs, performance issues, feature requests and
build/installation issues on GitHub. tag:bug_template</em>

**System information**
- Have I written custom code (as opposed to using a stock example script provided in TensorFlow):
- OS Platform and Distribution (e.g., Linux Ubuntu 16.04):
- Mobile device (e.g. iPhone 8, Pixel 2, Samsung Galaxy) if the issue happens on mobile device:
- TensorFlow installed from (source or binary):
- TensorFlow version (use command below):
- Python version:
- Bazel version (if compiling from source):
- GCC/Compiler version (if compiling from source):
- CUDA/cuDNN version:
- GPU model and memory:

You can collect some of this information using our environment capture
[script](https://github.com/tensorflow/tensorflow/tree/master/tools/tf_env_collect.sh)
You can also obtain the TensorFlow version with:
1. TF 1.0: `python -c "import tensorflow as tf; print(tf.GIT_VERSION, tf.VERSION)"`
2. TF 2.0: `python -c "import tensorflow as tf; print(tf.version.GIT_VERSION, tf.version.VERSION)"`

**Describe the current behavior**

**Describe the expected behavior**

**Standalone code to reproduce the issue**
Provide a reproducible test case that is the bare minimum necessary to generate
the problem. If possible, please share a link to Colab/Jupyter/any notebook.

**Other info / logs** Include any logs or source code that would be helpful to
diagnose the problem. If including tracebacks, please include the full
traceback. Large logs and files should be attached.

**Comment by amahendrakar** [comment]
I was able to reproduce the issue with TF v2.2 and TF-nightly. Please find the gist of it [here](https://colab.research.google.com/gist/amahendrakar/385c4e588b00f47665899b95e261ff2d/41760.ipynb).

**Comment by amahendrakar** [comment]
@dathudeptrai,
Could you please check [this workaround](https://github.com/tensorflow/tensorflow/issues/27120#issuecomment-540071844) from a similar issue and let us know if it helps. Thanks!

**Comment by dathudeptrai** [comment]
@amahendrakar I agree that the native solution is create multiple forward functions for each condition.  But don't you think this solution is not good interms of coding style?. Especially when you want to apply very complicated training procedures such as GAN, etc. Is there any good solution to replace this native solution ?

**Comment by mdanatg** [comment]
I had a look, I think there are two issues highlighted in this bug:

1. calling `model(x, training=False)` and then `model(x, training=True)` leads to error. This looks like a bug, but it can be worked around by decorating `call` with `@tf.function`. Ideally, Keras should rebuild the graph when the `training` argument changes.
2. it's unclear whether changing the layer's trainable attribute has the expected effect - @tomerk will Keras rebuild the graph when the `trainable` attribute of various layers is being changed? I think that would be the expected behavior.

I think both cases should be handled without the need for extra workarounds, although they may require additional smarts inside Keras to handle that correctly. I do agree that users should not need to resort to building multiple versions of the same model.

**Comment by tomerk** [comment]
As far as I can tell this doesn't have anything to do with Keras specifically. (and it doesn't even use Keras's own tf.functioning logic in fit/evaluate)

tf.functions will only capture python state at the time when they are first traced, so changing any python attribute after the first time you call the `tf.function`'d `grad` won't do anything.
@mdanatg I suppose the tf.function guide might be insufficient, because it seems to only talk about python side effects w/in the tf.function ( https://www.tensorflow.org/guide/function#python_side_effects ) but doesn't mention that changes to python state that happen outside the function will only be reflected in new traces.

If you want the trainable state to be reflected w/in your tf.function, you need to make sure to use different traces when you want to set the model.trainable to True and when you wan't to set it to False. You could do this by making your `grad` method take a `trainable` python boolean argument, then set the model's state at the start of `grad` inside of the tf.function tracing. (Just make sure to reset trainable at the end of your trace)

**Comment by MarkDaoust** [comment]
From your notebook, this is the problem:

```
@tf.function
def grad(model,...):
  ...

grad(model, ...)

model = SmallModel()

grad(model, ...)
```
For now, if you need to call a `tf.function` on multiple objects (that initialize their variables in the first call) you need to create a separate `@tf.function` for each instance. Or define it as a method.

See: https://github.com/keras-team/keras-io/pull/334

```
dense1 = tf.keras.layers.Dense(3)
dense2 = tf.keras.layers.Dense(3)

@tf.function
def fun(dense, input):
  return dense(input)
  
fun(dense1, input)
fun(dense2, input)
```
```
ValueError: tf.function-decorated function tried to create variables on non-first call.
```

**Comment by sushreebarsa** [comment]
Was able to reproduce the issue in TF v2.5,please find the gist[ here](https://colab.research.google.com/gist/sushreebarsa/cdf0d9d892362efc329004c7405cde4d/untitled111.ipynb)..Thanks !

**Comment by kaijennissen** [comment]
@dathudeptrai Did you figure out a way to switch between training generator and discriminator based on an attribute? I'm working on a similar problem where I also want to train one network for a given number of steps and then switch between training the first and the second. A minimal example can be found [here](https://colab.research.google.com/drive/1qhz0bTGwYvwTSVxNr7WC7kCOgYXXNV0p?usp=sharing). Any suggestions?

**Comment by dathudeptrai** [comment]
@kaijennissen you need use self.optimizer.iterations % 2 == 0 for if condition for example. Do not use any python scalar, boolean :D.

**Comment by kaijennissen** [comment]
@dathudeptrai That was fast, thanks!

**Comment by kaijennissen** [comment]
@dathudeptrai Unfortunately that doesn't work. Using f.e. `if self.w_optimizer.iterations < 50:` yields the following error message 
```OperatorNotAllowedInGraphError: using a `tf.Tensor` as a Python `bool` is not allowed: AutoGraph did convert this function. This might indicate you are trying to use an unsupported feature.```
This is strange because from my understand of the [docs](https://github.com/tensorflow/tensorflow/blob/master/tensorflow/python/autograph/g3doc/reference/control_flow.md#if-statements)  this should be converted.

**Comment by dathudeptrai** [comment]
@kaijennissen you need wrap `train_step` by @tf.function or use tf.cond rather than if. BTW, you need to use `Tensor` to perform if condition, do not use any python object in the if condition. For example, if A <= 5, A must be a tensor object such as optimizer.iterations.

```
@tf.function
def train_step(self, data):
....
```
